meta/src/ast.rs

```diff
@@ -58,15 +58,25 @@ pub enum Expr {
     /// Continues to match expressions until one of the strings in the `Vec` is found
     Skip(Vec<String>),
     /// Matches an expression and pushes it to the stack, e.g. `push(e)`
-    Push(Box<Expr>)
+    Push(Box<Expr>),
+
+    /// Internal Exprs
+
+    /// If the wrapped expression errors, return the parser state to where it was prior to
+    /// running the closure
+    RestoreOnErr(Box<Expr>)
```

Comment:
I'm still not quite sure how the AST nodes should be dealt with. Since there are two types, I don't know if the internal compiler types (skip, restore, etc.) should be visible here or what we should do with them.

Reply:
This is a great question. I'm not entirely sure myself. I'm of 2 minds. On the one hand, it's incredibly useful to have the internal compiler types here since they will necessarily be a product of the optimizer. On the other hand, I'm not entirely sure they should be exposed directly since I don't think we'll have any plans of making any explicit syntax for restore.